pkg/net/libp2p/libp2ptransport.go

```diff
@@ -0,0 +1,283 @@
+	wg.Add(len(t.membership) - 1)
+
+	for nodeID, nodeAddr := range t.membership {
+		if nodeID == t.ownID {
+			continue
+		}
+
+		go func(nodeID types.NodeID, nodeAddr string) {
+			defer wg.Done()
+
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s is connecting to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+			maddr, err := multiaddr.NewMultiaddr(nodeAddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse %v for node %v: %v", nodeAddr, nodeID, err))
+				return
+			}
+
+			// Extract the peer ID from the multiaddr.
+			info, err := peer.AddrInfoFromP2pAddr(maddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse addr %v for node %v: %v", maddr, nodeID, err))
+				return
+			}
+
+			t.host.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+			s, err := t.openStream(ctx, info.ID)
+			if err != nil {
+				t.logger.Log(logging.LevelError, fmt.Sprintf("couldn't open stream: %v", err))
+				return
+			}
+			t.addOutboundStream(nodeID, s)
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+		}(nodeID, nodeAddr)
+	}
+
+	wg.Wait()
+}
+
+func (t *Transport) openStream(ctx context.Context, p peer.ID) (network.Stream, error) {
+	timeout := 500 * time.Millisecond
+	for {
+		sctx, cancel := context.WithTimeout(ctx, timeout)
+		s, err := t.host.NewStream(sctx, p, ID)
+		cancel()
+		if err != nil {
+			t.logger.Log(logging.LevelDebug, "opening stream", "err", err.Error())
+			select {
+			case <-time.After(timeout):
+				continue
+			case <-ctx.Done():
+				return nil, fmt.Errorf("context closed")
+			}
+		}
+		return s, nil
+	}
+}
+
+func (t *Transport) Send(dest types.NodeID, payload *messagepb.Message) error {
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	t.outboundStreamsMx.Lock()
+	s, ok := t.outboundStreams[dest]
+	t.outboundStreamsMx.Unlock()
+	if !ok {
+		return fmt.Errorf("failed to get stream for node %v", dest)
+	}
+
+	w := bufio.NewWriter(bufio.NewWriter(s))
+
+	msg := TransportMessage{
+		t.ownID.Pb(),
+		outBytes,
+	}
+
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return err
+	}
+
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		w.Flush()
+	}
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+func (t *Transport) mirHandler(s network.Stream) {
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("mir handler for %s started", s.ID()))
+	defer t.logger.Log(logging.LevelDebug, fmt.Sprintf("mir handler for %s stopped", s.ID()))
+
+	defer s.Reset() // nolint
+
+	for {
+		var msg TransportMessage
+		err := msg.UnmarshalCBOR(s)
+		if err != nil {
+			t.logger.Log(logging.LevelError, "failed to read mir transport request", "err", err)
+			return
+		}
+
+		var payload messagepb.Message
+
+		if err := proto.Unmarshal(msg.Payload, &payload); err != nil {
```

Comment:
I know it is due to the message format used in Mir and the fact that protobuf can't be sent directly sent through libp2p, but we should maybe consider in the future using exclusively CBOR

Reply:
Yes, the whole event and message serialization could use some cleaning up. I've created an issue for that: #129